cunumeric/vectorize.py

```diff
@@ -0,0 +1,643 @@
+                a_tmp_base = a_tmp.base
+                task.add_input(a_tmp_base)
+                task.add_output(a_tmp_base)
+                task.add_alignment(first_array.base, a_tmp_base)
+
+        # add array arguments with read-only permissions
+        if num_args > 0:
+            start = 0
+            if first_array is None:
+                first_array = runtime.to_deferred_array(self._args[0]._thunk)
+                task.add_input(first_array.base)
+                start = 1
+            for i in range(start, num_args):
+                a_tmp = runtime.to_deferred_array(self._args[i]._thunk)
+                a_tmp_base = a_tmp.base
+                task.add_input(a_tmp_base)
+                task.add_alignment(first_array.base, a_tmp_base)
+
+        if is_gpu:
+            ptx_hash = hash(self._gpu_func[0])
+            task.add_scalar_arg(ptx_hash, ty.int64)
+        else:
+            task.add_scalar_arg(
+                self._cpu_func.address, ty.uint64
+            )  # type : ignore
+        task.execute()
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Union[Any, Tuple[Any]]:
+        # each time we call `vectorize` on a pyfunc we need to clear
+        # these lists to support different types of arguments passed
+        self._scalar_args.clear()
+        self._scalar_idxs.clear()
+        self._args.clear()
+        self._arg_names.clear()
+        self._scalar_names.clear()
+
+        scalar_idx = 0
+        for i, arg in enumerate(args):
+            if arg is None:
+                raise ValueError(
+                    "None is not supported in user function "
+                    "passed to cunumeric.vectorize"
+                )
+            elif np.ndim(arg) == 0:
+                if self._cache and not self._created:
+                    self._cached_scalar_types.append(type(arg))
+                elif self._cache:
+                    if self._cached_scalar_types[scalar_idx] != type(arg):
+                        raise TypeError(
+                            " Input arguments to vectorized function should"
+                            " have consistent types for each invocation"
+                        )
+                self._scalar_args.append(arg)
+                self._scalar_idxs.append(i)
+                scalar_idx += 1
+            else:
+                self._args.append(convert_to_cunumeric_ndarray(arg))
+
+        # first fill arrays to argnames, then scalars:
+        for i, k in enumerate(inspect.signature(self._pyfunc).parameters):
+            if not (i in self._scalar_idxs):
+                self._arg_names.append(k)
+
+        for i, k in enumerate(inspect.signature(self._pyfunc).parameters):
+            if i in self._scalar_idxs:
+                self._scalar_names.append(k)
+
+        self._kwargs = list(kwargs)
+        if len(self._kwargs) > 0:
+            raise NotImplementedError(
+                "kwargs are not supported in user functions"
+            )
```

Comment:
That said if `self._kwargs` is not even used, why store it at all? just check the length of `kwargs` is zero and bail if not